File: ObjC/CircularSegmentedControl/CircularSegmentedControl/ArcTextView.m
//
//  ArcTextView.m
//  CircularSegmentedControl
//
//  Created by Don Mag on 12/11/24.
//

#import "ArcTextView.h"

@implementation ArcTextView

- (instancetype)initWithFrame:(CGRect)frame {
	self = [super initWithFrame:frame];
	if (self) {
		[self commonInit];
	}
	return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
	self = [super initWithCoder:coder];
	if (self) {
		[self commonInit];
	}
	return self;
}

- (void)commonInit {
	self.backgroundColor = [UIColor clearColor];
	self.text = @"Text Along Arc";
	self.startAngle = 0.0;
	self.radius = 100.0;
	self.font = [UIFont systemFontOfSize:16];
	self.textColor = [UIColor blackColor];
}

- (void)setRadius:(CGFloat)radius {
	_radius = radius;
	[self setNeedsDisplay];
}

- (void)setFont:(UIFont *)font {
	_font = font;
	[self setNeedsDisplay];
}

- (void)setTextColor:(UIColor *)textColor {
	_textColor = textColor;
	[self setNeedsDisplay];
}

- (void)drawRect:(CGRect)rect {
	[super drawRect:rect];
	
	CGContextRef context = UIGraphicsGetCurrentContext();
	if (!context) return;
	
	// Center of the arc
	CGPoint center = CGPointMake(CGRectGetMidX(rect), CGRectGetMidY(rect));
	
	// Attributes for the text
	NSDictionary *attributes = @{
		NSFontAttributeName: self.font,
		NSForegroundColorAttributeName: self.textColor
	};
	
	// Total arc length
	CGFloat totalArcLength = 0;
	for (NSUInteger i = 0; i < self.text.length; i++) {
		NSString *charString = [self.text substringWithRange:NSMakeRange(i, 1)];
		CGSize charSize = [charString sizeWithAttributes:attributes];
		totalArcLength += charSize.width;
	}
	
	CGFloat totalArcAngle = totalArcLength / self.radius;
	
	// Adjust starting angle to center the text
	CGFloat currentAngle = self.startAngle - totalArcAngle / 2;
	
	// Draw each character
	for (NSUInteger i = 0; i < self.text.length; i++) {
		NSString *charString = [self.text substringWithRange:NSMakeRange(i, 1)];
		CGSize charSize = [charString sizeWithAttributes:attributes];
		
		// Calculate the position for the character along the arc
		CGFloat halfCharAngle = (charSize.width / self.radius) / 2;
		CGFloat charAngle = currentAngle + halfCharAngle;
		
		// Calculate the character's position
		CGFloat x = center.x + self.radius * cos(charAngle);
		CGFloat y = center.y + self.radius * sin(charAngle);
		
		// Save the context's state
		CGContextSaveGState(context);

		// Move to the character's position and rotate the context
		CGContextTranslateCTM(context, x, y);
		// Rotate to align with the arc
		CGContextRotateCTM(context, charAngle + M_PI_2);
		
		// Draw the character
		[charString drawAtPoint:CGPointMake(-charSize.width / 2, -charSize.height / 2) withAttributes:attributes];
		
		// Restore the context's state
		CGContextRestoreGState(context);
		
		// Update current angle
		currentAngle += charSize.width / self.radius;
	}
}

@end

File: Swift/CircularSegmentedControl/CircularSegmentedControl/ArcTextView.swift
//
//  ArcTextView.swift
//  CircularSegmentedControl
//
//  Created by Don Mag on 12/11/24.
//

import UIKit

class ArcTextView: UIView {
	var text: String = "Text Along Arc"
	var startAngle: CGFloat = 0.0
	var radius: CGFloat = 100.0 { didSet { setNeedsDisplay() } }
	
	var font: UIFont = .systemFont(ofSize: 16) { didSet { setNeedsDisplay() } }
	var textColor: UIColor = .black { didSet { setNeedsDisplay() } }
	
	override init(frame: CGRect) {
		super.init(frame: frame)
		commonInit()
	}
	required init?(coder: NSCoder) {
		super.init(coder: coder)
		commonInit()
	}
	private func commonInit() {
		backgroundColor = .clear
	}
	override func draw(_ rect: CGRect) {
		super.draw(rect)
		
		guard let context = UIGraphicsGetCurrentContext() else { return }
		
		// Center of the arc
		let center = CGPoint(x: rect.midX, y: rect.midY)
		
		// Attributes for the text
		let attributes: [NSAttributedString.Key: Any] = [
			.font: font,
			.foregroundColor: textColor,
		]
		// Total arc angle needed to draw the text
		var totalArcLength: CGFloat = 0
		for char in text {
			let charSize = (String(char) as NSString).size(withAttributes: attributes)
			totalArcLength += charSize.width
		}
		
		let totalArcAngle = totalArcLength / radius
		
		// Adjust starting angle to center the text if necessary
		var currentAngle = startAngle - totalArcAngle / 2
		
		// Draw each character
		for char in text {
			let charString = String(char)
			let charSize = (charString as NSString).size(withAttributes: attributes)
			
			// Calculate the position for the character along the arc
			let halfCharAngle = (charSize.width / radius) / 2
			let charAngle = currentAngle + halfCharAngle
			
			// Calculate the character's position
			let x = center.x + radius * cos(charAngle)
			let y = center.y + radius * sin(charAngle)
			
			// Save the context's state
			context.saveGState()
			
			// Move to the character's position and rotate the context
			context.translateBy(x: x, y: y)
			// Rotate to align with the arc
			context.rotate(by: charAngle + .pi / 2)
			
			// Draw the character
			charString.draw(at: CGPoint(x: -charSize.width / 2, y: -charSize.height / 2), withAttributes: attributes)
			
			// Restore the context's state
			context.restoreGState()
			
			// Update the current angle
			currentAngle += charSize.width / radius
		}
	}
}
